Added font bold when isDefaultAction is true in CupertinoDialogAction

riccardoratta opened the pull request:
## Description

From the `isDefaultAction` doc in the `CupertinoDialogAction` class.

https://github.com/flutter/flutter/blob/eaf058d81d868b8570ef75dba8bcbfd0c0149080/packages/flutter/lib/src/cupertino/dialog.dart#L1045-L1048

But actually changing this parameter doesn't cause any effect.

### Example

By launching this app

```dart
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key}) : super(key: key);

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return CupertinoPageScaffold(
      child: Center(
        child: CupertinoButton(
          child: Text('Alert'),
          onPressed: () {
            showDialog<void>(
                context: context,
                builder: (BuildContext context) {
                  return CupertinoAlertDialog(
                    title: Text('Title'),
                    content: Text('Content'),
                    actions: <Widget>[
                      CupertinoDialogAction(
                        child: Text('Cancel'),
                        onPressed: () => Navigator.of(context).pop(),
                      ),
                      CupertinoDialogAction(
                        child: Text('Ok'),
                        onPressed: () => Navigator.of(context).pop(),
                        isDefaultAction: true,
                      ),
                    ],
                  );
                });
          },
        ),
      ),
    );
  }
}
```

you will get

<img width="492" alt="Screenshot 2019-04-19 at 15 17 39" src="https://user-images.githubusercontent.com/6663880/56426439-9ed87080-62b8-11e9-92c0-227028a071f3.png">

where the "ok" button is not bold.

### Solution

The change seem to solve the problem.

<img width="492" alt="Screenshot 2019-04-19 at 15 27 09" src="https://user-images.githubusercontent.com/6663880/56426541-f676dc00-62b8-11e9-8657-7642ec6fcfe8.png">

Also `isDefaultAction` variable was never used in the code.

## Related Issues

No related issues found.

## Tests

Only performed visual test on the application (see screenshots above).

## Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

riccardoratta committed:
Added font change when isDefaultAction is true

LongCatIsLooong commented on packages/flutter/lib/src/cupertino/dialog.dart:
```diff
@@ -1140,6 +1140,10 @@ class CupertinoDialogAction extends StatelessWidget {
     TextStyle style = _kCupertinoDialogActionStyle;
     style = style.merge(textStyle);
 
+    if (isDefaultAction) {
```
It appears that `isDefaultAction` can be explicitly set to `null` in the constructor. Could you add an assert to `CupertinoDialogAction`'s constructor to make sure `isDefaultAction` is not nullable?

LongCatIsLooong commented:
Thanks for the pull request @riccardoratta! 
Could you add a few tests in `packages/flutter/test/cupertino/dialog_test.dart` to make sure `isDefaultAction` works? The "Dialog destructive action styles" test in the same file looks like a good example in case you're looking for one.

LongCatIsLooong commented:
`UIAlertController.preferredAction` does the same on iOS but for the same `UIAlertController` you can have at most one preferred action. I think it makes sense to enforce the same behavior in the constructor of `CupertinoAlertDialog`.


Also the name `isDefaultAction` might be a bit misleading for iOS developers since in the constructor it's juxtaposed with `isDestructiveAction`, it's easily mistakable for [UIAlertAction.Style.default](https://developer.apple.com/documentation/uikit/uialertaction/style).

CC @xster @matthew-carroll

matthew-carroll commented:
Generally, LGTM.

@LongCatIsLooong  Changing the name would be a breaking change. Is it significant enough of an issue to warrant that?

riccardoratta committed:
Check if isDefaultAction and isDestructiveAction are not null

riccardoratta committed:
Check if default and destructive style are true at the same time

riccardoratta committed:
Fixed test for destructive dialog action

Removed `isDefaultAction` option since is not valid anymore (due to a new assert). Also performed a check on the final color of the text, that must be destructiveRed.

riccardoratta committed:
Fixed test for default dialog action

matthew-carroll commented on packages/flutter/lib/src/cupertino/dialog.dart:
```diff
@@ -1034,7 +1034,10 @@ class CupertinoDialogAction extends StatelessWidget {
     @required this.child,
-  }) : assert(child != null);
+  }) : assert(child != null),
+       assert(isDefaultAction != null || isDestructiveAction != null),
```
Is this condition correct? I don't think that a button needs to be default or destructive - consider an alert dialog with 5 buttons.

riccardoratta replied:
It's just a check to prevent the user to explicitly set `null` the constructor, their default value is `false`.

Otherwise these conditions will throw an exception:

https://github.com/flutter/flutter/blob/cda5dfbf03ca6534b47ddd1b23c20b54a68a2fe1/packages/flutter/lib/src/cupertino/dialog.dart#L1146-L1152

matthew-carroll replied:
Ok, if the goal is to ensure that the values are non-null, then shouldn't we have 2 asserts? 1 for each property?

```dart
assert(isDefaultAction != null),
assert(isDestructiveAction != null),
// ...
```

riccardoratta replied:
You're right! Fixed, thanks.

matthew-carroll commented on packages/flutter/test/cupertino/dialog_test.dart:
```diff
@@ -132,28 +132,26 @@ void main() {
+    expect(widget.style.fontWeight, equals(FontWeight.bold));
   });
 
-  testWidgets('Default and destructive style', (WidgetTester tester) async {
```
Wasn't this a legitimate test? If so, we should add a new test rather than rewrite an existing one.

riccardoratta replied:
Due to the new assertion I can't set both parameters (`isDefaultAction`, `isDestructiveAction`) to true, the compiler warn: "Evaluation of this constant expression throws an exception. `dart(const_eval_throws_exception)`"

matthew-carroll replied:
Ok, makes sense. Thanks.

LongCatIsLooong replied:
`isDefaultAction` and `isDestructiveAction` seems to be orthogonal. On iOS an action can be both default(`preferredAction`) and destructive.

matthew-carroll commented on packages/flutter/test/cupertino/dialog_test.dart:
```diff
@@ -132,28 +132,26 @@ void main() {
 
-    expect(widget.style.fontWeight, equals(FontWeight.w400));
-    expect(widget.style.color.red, greaterThan(widget.style.color.blue));
+    expect(widget.style.color.withOpacity(1), equals(CupertinoColors.destructiveRed));
```
Shouldn't the `widget.style.color` be exactly the same as `CupertinoColors.destructiveRed`? Or, if the widget is internally reducing the opacity, then can we match something like `CupertinoColors.destructiveRed.withOpacity(...)`

riccardoratta replied:
Since `onPressed` is not set the action button is not `enabled` and the text opacity is reduced.

https://github.com/flutter/flutter/blob/cda5dfbf03ca6534b47ddd1b23c20b54a68a2fe1/packages/flutter/lib/src/cupertino/dialog.dart#L1154-L1156

If you prefer I can flip the test to `expect(widget.style.color, equals(CupertinoColors.destructiveRed.withOpacity(0.5)))`, also, if it's possible, can be a good idea add a comment explaining why the opacity of the destructiveRed is 50% in the test.

matthew-carroll replied:
Yeah, I think it would be good to explicitly use the actual opacity as you suggested.

Also, if such tests do not exist, can you add 1 or 2 tests that test the opacity of the destructive button when it's enabled and when it's disabled?

riccardoratta replied:
@matthew-carroll I've tried to add a specify test for the opacity when the dialog action is disabled, but the opacity doesn't correspond to `0.5`. Can't figure out why.

matthew-carroll replied:
It looked like there was a followup comment to this about not being able to get opacity working in the test. I'm not sure why GitHub isn't showing me that comment. If opacity isn't working as expected then I'd recommend adding a TODO stating the problem and requesting it be fixed. Then we can take care of it later.

Does that sound good to you @xster and @LongCatIsLooong ?

riccardoratta replied:
@matthew-carroll I deleted it, the opacity that I get from the color isn’t exact 0.5 because its encoded in the range [0,255], but it’s a normal behavior.

riccardoratta committed:
Splitted assert

LongCatIsLooong commented on packages/flutter/lib/src/cupertino/dialog.dart:
```diff
@@ -1034,7 +1034,11 @@ class CupertinoDialogAction extends StatelessWidget {
+       assert(isDefaultAction != null),
+       assert(isDestructiveAction != null),
+       assert(!(isDefaultAction && isDestructiveAction),
+          "Action can't be default and destructive at the same time");
```
On iOS an action can be both default(preferredAction) and destructive.

riccardoratta replied:
Thanks to point it out, I believed was not possible since it's an enum on iOS.

riccardoratta replied:
If I'm not wrong on iOS you have en enum for `default`, `cancel` and `destructive` and a boolean for `preferredAction `?. Maybe I can bring the behavior closer to the native one, but it will be a breaking change.

LongCatIsLooong replied:
Sorry for the ambiguity. The iOS enum case [`default`](https://developer.apple.com/documentation/uikit/uialertaction/style/default) doesn't make the action button bold. `isDefaultAction` is similar to [UIAlertController.preferredaction](https://developer.apple.com/documentation/uikit/uialertcontroller/1620102-preferredaction) which takes a `UIAlertAction?` and highlights it.

I'm filing a new issue for the misleading parameter name.

riccardoratta committed:
Removed check for default and destructive that must be both not true

xster commented on packages/flutter/lib/src/cupertino/dialog.dart:
```diff
@@ -1140,6 +1142,10 @@ class CupertinoDialogAction extends StatelessWidget {
     style = style.merge(textStyle);
 
+    if (isDefaultAction) {
+      style = style.copyWith(fontWeight: FontWeight.bold);
```
This isn't right. The font weight of an iOS preferred action is w600/semi-bold according to https://developer.apple.com/design/resources. Bold is w700.

riccardoratta replied:
@xster Thanks! Didn't noticed that.

<img width="651" alt="Screenshot 2019-05-03 at 11 43 58" src="https://user-images.githubusercontent.com/6663880/57130232-21712d80-6d99-11e9-8e60-58a8fd61de95.png">

riccardoratta committed:
Changed font weight from bold to w600 (semibold)

This is the correct weight for _preferred_ action (here called default).

riccardoratta committed:
Modified test to check exactly if the text color is `destructiveRed`

The alpha channel of the tested color is _removed_ since the scope of this test is only to check the output color when `isDestructiveAction` is `true`. Other tests will be performed to check the opacity (or alpha) when the action is enabled or not.

riccardoratta committed:
Updated font weight when the style is default

riccardoratta committed:
Revert test to original one with both default and destructive styles

This test was wrongly removed by me because was not possible anymore to set both style properties to `true`.

riccardoratta commented:
Don't know why this test doesn't work
```dart
testWidgets('Dialog disabled action style', (WidgetTester tester) async {
  await tester.pumpWidget(boilerplate(const CupertinoDialogAction(
    child: Text('Ok'),
  )));

  final DefaultTextStyle widget = tester.widget(find.byType(DefaultTextStyle));

  expect(widget.style.color.opacity, equals(0.5));
});
```

```
The following TestFailure object was thrown running a test:
  Expected: <0.5>
  Actual: <0.5019607843137255>
```

riccardoratta committed:
Added test to check text opacity when dialog action is not enabled

xster commented on packages/flutter/lib/src/cupertino/dialog.dart:
```diff
@@ -1034,7 +1034,9 @@ class CupertinoDialogAction extends StatelessWidget {
-  }) : assert(child != null);
+  }) : assert(child != null),
+       assert(isDefaultAction != null),
+       assert(isDestructiveAction != null);
```
On lines 1049 and 1054, add more lines describing their defaulting behavior:

`Defaults to false and cannot be null`

xster reviewed (approved):
If the tests were only testing the color in disabled state, we should add tests for enabled states too.

Otherwise LGTM.

riccardoratta committed:
Added test for dialog enabled action style

riccardoratta committed:
Added doc for describe default behavior of default and destructive styles

LongCatIsLooong commented on packages/flutter/lib/src/cupertino/dialog.dart:
```diff
@@ -1046,12 +1046,13 @@ class CupertinoDialogAction extends StatelessWidget {
   /// Whether this action destroys an object.
   ///
-  /// For example, an action that deletes an email is destructive.
+  /// For example, an action that deletes an email is destructive. Defaults to 
```
Please remove trailing white space.

riccardoratta committed:
Removed trailing white space

riccardoratta committed:
Added point at the end of a the sentence

xster committed:
Small update to the doc
